File: leetcode/0437-path-sum-iii/solution.cpp
#include <unordered_map>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        unordered_map<long long, int> prefixCounts;
        prefixCounts[0] = 1;
        int count = 0;
        dfs(root, 0LL, targetSum, prefixCounts, count);
        return count;
    }

private:
    void dfs(TreeNode* node, long long runningSum, int targetSum,
             unordered_map<long long, int>& prefixCounts, int& count) {
        if (node == nullptr) {
            return;
        }
        runningSum += node->val;
        auto it = prefixCounts.find(runningSum - targetSum);
        if (it != prefixCounts.end()) {
            count += it->second;
        }
        prefixCounts[runningSum]++;
        dfs(node->left, runningSum, targetSum, prefixCounts, count);
        dfs(node->right, runningSum, targetSum, prefixCounts, count);
        prefixCounts[runningSum]--;
    }
};

File: leetcode/0437-path-sum-iii/Solution.java
import java.util.HashMap;
import java.util.Map;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int targetSum;
    private int count;
    private Map<Long, Integer> prefixCounts;

    public int pathSum(TreeNode root, int targetSum) {
        this.targetSum = targetSum;
        this.count = 0;
        this.prefixCounts = new HashMap<>();
        prefixCounts.put(0L, 1);
        dfs(root, 0L);
        return count;
    }

    private void dfs(TreeNode node, long runningSum) {
        if (node == null) {
            return;
        }
        runningSum += node.val;
        count += prefixCounts.getOrDefault(runningSum - targetSum, 0);
        prefixCounts.merge(runningSum, 1, Integer::sum);
        dfs(node.left, runningSum);
        dfs(node.right, runningSum);
        prefixCounts.merge(runningSum, -1, Integer::sum);
    }
}
